RFC 1012: Range types (#57)
edgedb/rfcs: text/1012-range-types.rst

  function range(
      lower: optional anypoint = {},
      upper: optional anypoint = {},
      named only inclow: bool = true,
      named only inchigh: bool = false
  ) -> range<anypoint>

The default values are chosen so as to minimize the amount of keystrokes
required to define ranges for common use-cases such as for iterators.

The ``inclow`` and ``inchigh`` arguments are ``named only`` because when their
default values need to be overridden it is entirely plausible that only one of
them needs to be specified and thus using positional arguments would be
awkward.

A ``multirange`` could be constructed using a function that takes a set of
basic ``range`` values::

  function multirange(
      ranges: set of range<anypoint>
  ) -> multirange<anypoint>

The constructor must normalize the ranges to be non-overlapping and order them
based on boundaries. This will ensure that when a ``multirange`` is unpacked
into its component ranges, the order of the ``range`` values would be based on
their boundaries rather than arbitrary.

Accessing range components
--------------------------

Several functions will be used to access ``range`` components::

>   function get_range_upper(range<anypoint>) -> optional anypoint
>   function get_range_lower(range<anypoint>) -> optional anypoint
>   function is_inclusive_lower(range<anypoint>) -> bool
>   function is_inclusive_upper(range<anypoint>) -> bool

Reviewer: ```suggestion
  function range_get_upper(range<anypoint>) -> optional anypoint
  function range_get_lower(range<anypoint>) -> optional anypoint
  function range_is_inclusive_lower(range<anypoint>) -> bool
  function range_is_inclusive_upper(range<anypoint>) -> bool
```

Author: Yeah, good point about our naming scheme.

Accessing multirange components
-------------------------------

There will be a function for extracting sub-ranges from a multirange::

>   function get_subranges(multirange<anypoint>) -> set of range<anypoint>

Reviewer: Maybe

```suggestion
  function multirange_unpack(multirange<anypoint>) -> set of range<anypoint>
```

?

Author: This is very long name... maybe `to_range(multirange<anypoint>)` instead? It's a conversion function of sorts.

Reviewer: `to_range` implies that a single range would be returned, `to_ranges` would be better.

Author: Fair enough.

By default this function will return the ranges ordered by boundaries

> fromlowest to highest. This is the order of definition and Postgres by default

Reviewer: ```suggestion
from lowest to highest. This is the order of definition and Postgres by default
```

Author: Fixed typo.

returns the sub-intervals in that order.

Helper functions
----------------

In addition to the functions mentioned above there must be a way to generate
an ordered set of values from a ``range`` or ``multirange``::

  function range_unpack(
      val: range<anydiscrete>,

>       step: anydiscrete = DEFAULT_STEP

Reviewer: Would be easier to add overloads instead of sentinel `DEFAULT_STEP`.

Author: Yes, you're right, I was sloppy here and I should explain that we want overloads.

  ) -> set of anydiscrete

  function range_unpack(
      val: multirange<anydiscrete>,
      step: anydiscrete = DEFAULT_STEP
  ) -> set of anydiscrete

  function range_unpack(
      val: range<anycontiguous>,

>       step: anycontiguous

Reviewer: We would have to invent another type: `increment<anycontiguous>` (and `increment<anydiscrete>`), which would internally resolve into the appropriate increment type of a given point type, i.e. `duration` for `datetime`.

Author: I don't think that inventing another type makes any sense here because for `local_date`, `local_time`, and `datetime` the step is `duration`, but it's also a valid step for `duration` itself. Trying to link them by marking them with abstract types doesn't seem to work because abstract types don't let us see the valid pairs.

Date and time scalars seem to be the only current exception, but if we ever have another one, generalizing this mechanism would be near impossible. On the other hand, encoding this feature as a particular overload for `range_unpack` is clear and future-proof.

Reviewer: `increment<type>` is not "another type", it's an alias for the return type of operator `-(type, type)`, i.e. it needs no actual formal implementation aside from a bit of schema magic.

Author: Wait, are you suggesting that `increment<type>` is a type expression that isn't independently defined, but instead literally looks up the operator `-(type, type)` and returns the declared return type of that operator?

And the main reason for this thing even existing is to be used in formal generic function definitions like `range_unpack(val: range<anypoint>, step: increment<anypoint>)`? What's the benefit? We would still implement the specific overloaded versions for each specific range type, I think.

If it is independently defined, then it would place the burden of reconciling it with the operator, so maybe not a great option.

Reviewer: > but instead literally looks up the operator `-(type, type)` and returns the declared return type of that operator?

This.

> We would still implement the specific overloaded versions for each specific range type, I think.

If we implement the "magic" type then we wouldn't necessarily need to implement the overloads, because the compiler can just assume that all descendants of `anypoint` implement the necessary generator.

Author: Added `increment<type>` to the RFC.

  ) -> set of anycontiguous

  function range_unpack(
      val: multirange<anycontiguous>,
      step: anycontiguous
  ) -> set of anycontiguous

The ``range_unpack`` function for ``anydiscrete`` intervals **must** have a
default (minimal) value for the ``step``. On the other hand ``anycontiguous``
intervals **must** specify the ``step`` explicitly in all cases as no
"natural" step value can be asumed. When invoked on an unbounded interval
``range_unpack`` will raise an error.

We also need functions for determining overlapping and whether something is
contained in a range::

  function contains(
      haystack: range<anypoint>,
      needle: range<anypoint>
  ) -> bool

  function contains(
      haystack: range<anypoint>,
      needle: anypoint
  ) -> bool

  function contains(
      haystack: multirange<anypoint>,
      needle: multirange<anypoint>
  ) -> bool

  function contains(
      haystack: multirange<anypoint>,
      needle: anypoint
  ) -> bool

  function overlaps(
      haystack: range<anypoint>,
      needle: range<anypoint>
  ) -> bool

  function overlaps(
      haystack: multirange<anypoint>,
      needle: multirange<anypoint>
  ) -> bool

> It should be noted that if ``contains(A, B) = true`` then ``overlaps(A, B) =

Reviewer: Postgres defines a bunch more functions for ranges, such as relative disposition (e.g. "strictly to the left of", "strictly to the right of", etc.).  I think we should reflect those as well.

Author: Hmm, yeah I missed an important bit here. We need to define `>`, `>=`, `<`, and `<=` for everything in EdgeQL (because everything must be orderable). So I'll add that to "operators". Meanwhile we'll have functions for "strictly to the left/right" and "does not extend to the left/right of".

The question is what to call these? They are wordy and I'd like to avoid full sentences as function names.

true``.

Operators
---------

The operators ``+`` and ``-`` should be defined for various combinations of
``range`` and ``multirange``. The semantics of ``+`` is a "union", whereas the

> ``-`` allows excluding some values from an interval. Both of these operators

Reviewer: What about the intersection of ranges?

Author: Yeah, you're right.
